### PR TITLE
#141 - Added Header.Wrap

### LINKED ISSUE #141: Missing abstract decorator for `Header`
There is no abstract decorator for building custom headers, it would be nice to have Header.Wrap to create new header classes by composition

### PR DESCRIPTION
**Pull request by olegmoz**
Closes #141

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#142](https://github.com/artipie/http/issues/142) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/db4f89f2-b958-409b-b1b3-d8f7f82b7444); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artipie/http/rs/Header.java [hunk 1]
```diff
@@ -112,4 +112,56 @@ public String toString() {
...
+        }
+
+        @Override
+        public String getValue() {
+            return this.header.getValue();
+        }
+
+        @Override
+        public String setValue(final String value) {
+            return this.header.setValue(value);
```
Comment (Vatavuk) [inline src/main/java/com/artipie/http/rs/Header.java:149]: @olegmoz let's throw exception here to make this class `immutable`

File: src/main/java/com/artipie/http/rs/Header.java [hunk 2]
```diff
@@ -112,4 +112,56 @@ public String toString() {
...
+         * Ctor.
+         *
+         * @param header Header.
+         */
+        protected Wrap(final Map.Entry<String, String> header) {
+            this.header = header;
+        }
+
+        @Override
+        public String getKey() {
```
Comment (Vatavuk) [inline src/main/java/com/artipie/http/rs/Header.java:138]: @olegmoz all methods should be final

File: src/main/java/com/artipie/http/rs/Header.java [hunk 3]
```diff
@@ -86,7 +86,7 @@ public boolean equals(final Object that) {
         if (this == that) {
             return true;
         }
-        if (that == null || getClass() != that.getClass()) {
```
Comment (Vatavuk) [inline src/main/java/com/artipie/http/rs/Header.java:89]: @olegmoz why have you remove this check? If we compare this class with non-header object we will get class cast exception

File: src/main/java/com/artipie/http/rs/Header.java (hunk 1 above)
Comment (olegmoz) [reply src/main/java/com/artipie/http/rs/Header.java:149]: @Vatavuk I think that immutability should depend on `header` and this wrapper does not have to enforce it

File: src/main/java/com/artipie/http/rs/Header.java (hunk 2 above)
Comment (olegmoz) [reply src/main/java/com/artipie/http/rs/Header.java]: @Vatavuk fixed, thanks

File: src/main/java/com/artipie/http/rs/Header.java (hunk 3 above)
Comment (olegmoz) [reply src/main/java/com/artipie/http/rs/Header.java]: @Vatavuk thanks for pointing this out, fixed it and made proper `equals` in `Wrap` too

File: src/main/java/com/artipie/http/rs/Header.java (hunk 1 above)
Comment (Vatavuk) [reply src/main/java/com/artipie/http/rs/Header.java:149]: @olegmoz ok

**Comment by 0crat** [comment]
Job was finished in 4 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @vatavuk[/z](https://www.0crat.com/u/vatavuk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@0crat quality good

**Pull request merged**